### ISSUE TITLE
problems with 'this'

### ISSUE THREAD

**Issue by doremaxime** [ga-wdi-boston/game-project#469]
When I click on the board to place a token, an issue comes up highlighting the following from the setToken function. 
```javascript
this.innerHTML === ''
```
function:
```javascript
const setToken = function () {
  if ((count % 2 === 0) && (this.innerHTML === '')) {
    this.innerHTML = 'x';
    console.log('1');
    checkWinner();
    console.log('1.1');
    count++;
    console.log('1.2');
    user = 'o';
    console.log('1.3');
  } else if ((count % 2 !== 0) && (this.innerHTML === '')) {
    this.innerHTML = 'o';
    console.log(2);
    checkWinner();
    console.log('2.1');
    count++;
    console.log('2.2');
    user = 'x';
    console.log('2.3');
  } else {
    console.log('already taken');
  }
};

$('.circle').on('click', function () {
  $(this).closest('.circle').append(setToken());
  $(this).click(false);
});
```
if i take the line out of my code, it keeps processing but of course that it is not what i want. i would like it to be able to recognize if there is nothing in that circle's innerHTML so that it can go through the if/else.
the console.log statements are just for debugging.

**Comment by doremaxime** [comment]
still can't figure this... i need to find a way in the first condition that it can recognize that the innerHTML of the id that's been clicked is empty but am still not successful and eating me alive haha

**Comment by bernardlee** [comment]
Add a line, `console.log(this);` at the beginning of `setToken()` and at the beginning of the callback in your click handler for `circle` classes, to see if `this` is really what you think it should be and if it's the appropriate entity to be passed into another function as an argument.

**Comment by doremaxime** [comment]
well what i am trying to do is not use the 'this'. i am now putting 'gameBoard[index]' instead and it's not working. i believe that issue lies with the append of:
```
$('.circle').on('click', function () {
  $(this).closest('.circle').append(setToken());
  $(this).click(false);
});
```
but i can't make sense of it. 'setToken()' is being called so what the value it gets is what will be appended no?

updated function:
```
const setToken = function (index) {
  if ((count % 2 === 0) && (gameBoard[index] === '')) {
    gameBoard[index] = 'x';
    console.log('1');
    checkWinner();
    console.log('1.1');
    count++;
    console.log('1.2');
    user = 'o';
    console.log('1.3');
  } else if ((count % 2 !== 0) && (gameBoard[index] === '')) {
    gameBoard[index] = 'o';
    console.log('2');
    checkWinner();
    console.log('2.1');
    count++;
    console.log('2.2');
    user = 'x';
    console.log('2.3');
  } else {
    console.log('already taken');
  }
};
```

**Comment by bernardlee** [comment]
`setToken()` doesn't return anything so its value is going to be `undefined`.  In your click handler, the statement `$(this).closest('.circle').append(setToken());` is equivalent to `$(this).closest('.circle').append(undefined);`.

**Comment by doremaxime** [comment]
but if i do:
```
return gameBoard[index] = 'x';
```
it'll return it and won't do everything underneath.
it then needs to check for a winner, increase the count and change the user
(not thinking you are wrong, i'm explaining my thoughts which are wrong haha)

**Comment by doremaxime** [comment]
it then needs to check for a winner, increase the count and change the user

**Comment by bernardlee** [comment]
Correct.  `return` will stop the execution of the current function and return control back to its caller.  
One way that you can simplify and potentially solve this is to change the control flow.  Ideally, your functions should perform one particular task.  In this case, `setToken()` should just set a token.

As it is right now, you're telling `setToken()`, "Check the turn counter and check if this spot on the board is taken and then put a token in it if it's empty.  Also, check if there's a winner and increase the counter."  You're telling it to do more than it should.

Hint: Since you want to call `checkWinner()` whenever a token has been successfully placed, maybe it can be moved elsewhere.

**Comment by doremaxime** [comment]
well i've tried putting ```checkWinner()``` after ```  $(this).closest('.circle').append(setToken());
``` but that doesn't work either. I also tried putting it at the very end setToken function but that doesn't work either. both attempts makes the console log 'already taken' **twice** even though i clicked it only once. I believe ```checkWinner``` has to be checked after a token has been place.

**Comment by doremaxime** [comment]
I've tried so many variation. I feel like my function now is better:
```
const setToken = function (index) {
  if (gameBoard[index] === '') {
    gameBoard[index] = user;
    if (user === 'x') {
      user = 'o';
    } else if (user === 'o') {
      user = 'x';
    } else {
    console.log('already taken');
  }
}

  checkWinner();
};

$('.gameboard').on('click', function () {
  $(this).closest('.circle').text(user);
  $(this).click(false);
});
```
but nothing is appearing on the board.

index.js
```
const authEvents = require('./auth/events.js');

// On Document ready
$(() => {
  authEvents.addHandlers();
  $('#user-destroy').on('submit', authEvents.onDeleteUser);
  $('#edit-user').on('submit', authEvents.onPatchUser);
  $('#add-user').on('submit', authEvents.onPostUser);
  $('#TL').on('click', gameEngine.setToken);
  $('#TC').on('click', gameEngine.setToken);
  $('#TR').on('click', gameEngine.setToken);
  $('#ML').on('click', gameEngine.setToken);
  $('#MC').on('click', gameEngine.setToken);
  $('#MR').on('click', gameEngine.setToken);
  $('#BL').on('click', gameEngine.setToken);
  $('#BC').on('click', gameEngine.setToken);
  $('#BR').on('click', gameEngine.setToken);
  $('.restart').on('click', gameEngine.restart);
  $('.showInfo').on('click', gameEngine.showInfo);
});
```

**Comment by doremaxime** [comment]
how can i get Xs and Os to appear?

**Comment by bernardlee** [comment]
I'll clone your repository and take a look.

**Comment by bernardlee** [comment]
The reason that `setToken()` and, consequently, `checkWinner()` get called twice is because they're attached to two click handlers.  `.gameboard` has a click handler to call `setToken()` and each of the individual `<div>`s also have a click handler that calls `setToken()`.

`setToken()` is also defined with a parameter but none of the calls to it passes an argument in.

Also, the line `$(this).closest('.circle').text(user)` in your `.gameboard` click handler won't work because `this` is going to be the `gameboard` `<div>` and `.closest()` only traverses through the current element and its ancestors to find matches.

**Comment by bernardlee** [comment]
`setToken()` is also defined with a parameter but none of the calls to it passes an argument in.

**Comment by bernardlee** [comment]
Also, the line `$(this).closest('.circle').text(user)` in your `.gameboard` click handler won't work because `this` is going to be the `gameboard` `<div>` and `.closest()` only traverses through the current element and its ancestors to find matches.

**Comment by doremaxime** [comment]
Thank you very much!
Still looking at it and making adjustments.

**Comment by doremaxime** [comment]
hmmm... for the ```$(this).closest('.circle').text(user)``` i thought it would enter the text of whatever the ```user``` at that time would be to the closest circle that would be clicked on the board. ok, I commented that click function out. 
I add the indexes to the individual click functions in the index.js as such:
```
  $('#TL').on('click', gameEngine.setToken(0));
  $('#TC').on('click', gameEngine.setToken(1));
  $('#TR').on('click', gameEngine.setToken(2));
  $('#ML').on('click', gameEngine.setToken(3));
  $('#MC').on('click', gameEngine.setToken(4));
  $('#MR').on('click', gameEngine.setToken(5));
  $('#BL').on('click', gameEngine.setToken(6));
  $('#BC').on('click', gameEngine.setToken(7));
  $('#BR').on('click', gameEngine.setToken(8));
```
these will be used in as arguments for setToken's parameters.
But by having ```$().on('click',.....``` it only makes it so that WHEN you click, that function will called upon right? I ask because the console logs ```X Wins!``` as the page loads...

**Comment by doremaxime** [comment]
I understand that 'this' was not being what I thought it was. using console.log between different lines helps to find that out.
